[RFC] Add Secure partitions to OP-TEE image

jellesels-arm opened the pull request:
<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->
We need to be able to load Secure Partition (SPs). SPs will be loaded during the initialization of the kernel. This means that the SP images need to be embedded in the OP-TEE image. The process to achieve this is based around the early_tas. 

-ta_store is renamed to ts_store. This is to make it clear that the stores can be used by all Trusted Services (both TAs and SPs).
-An embedded_ts is created. This combines the shared logic of both early_ta's and embedded_sp's. 
-embedded_sp is create as the early_ta counterpart for loading the SPs.

jenswi-linaro commented on core/include/kernel/ts_store.h:
```diff
@@ -0,0 +1,75 @@
+	void (*close)(struct ts_store_handle *h);
+};
+
+struct ts_store_handle;
```
No need to repeat this, it's already defined above.

jellesels-arm replied:
Done

jenswi-linaro commented on core/pta/system.c:
```diff
@@ -226,7 +227,7 @@ static void ta_bin_close(void *ptr)
 }
 
-static TEE_Result system_open_ta_binary(struct system_ctx *ctx,
+static TEE_Result system_open_ts_binary(struct system_ctx *ctx,
```
I think we can keep the name as `system_open_ta_binary()` for now.

jellesels-arm replied:
done

jenswi-linaro commented on core/include/kernel/ts_store.h:
```diff
@@ -0,0 +1,75 @@
+ */
+#define TEE_TS_REGISTER_TS_STORE(prio) \
+	int __tee_ts_store_##prio __unused; \
+	SCATTERED_ARRAY_DEFINE_PG_ITEM_ORDERED(ts_stores, prio, \
```
It doesn't make sense to me to mix TA and SP stores. When a TA or an SP is requested the caller knows if it's expecting a TA or a SP. Given that we could let the SCATTERED_ARRAY* macros keep the two stores apart so they never get mixed. It still makes sense to have common "ts" structs and functions. So we'd have something like:
``` C
#define TEE_TS_REGISTER_TA_STORE(prio) \
￼       int __tee_ta_store_##prio __unused; \
￼       SCATTERED_ARRAY_DEFINE_PG_ITEM_ORDERED(ta_stores, prio, \
￼                                              struct ts_store_ops)
#define TEE_TS_REGISTER_SP_STORE(prio) \
￼       int __tee_sp_store_##prio __unused; \
￼       SCATTERED_ARRAY_DEFINE_PG_ITEM_ORDERED(sp_stores, prio, \
￼                                              struct ts_store_ops)
```
Does it make sense to you?

jellesels-arm replied:
I changed it here https://github.com/OP-TEE/optee_os/pull/4177/commits/cf2a539e91e1b5f29cf6bc225e7020a5243fee4d

jenswi-linaro reviewed (commented):
Comments for the commit "SPMC: Rename ta_store to ts_store"

Please change the title of the commit to something like:
`core: rename ta_store to ts_store`

jenswi-linaro commented:
Please rebase this on master

jenswi-linaro reviewed (commented):
The last commit "core: create separate stores for SPs and TAs" is really a fixup commit which should be merged into the the earlier commits as appropriate.

I'd rather go directly from `TEE_TA_REGISTER_TA_STORE` to `TEE_TS_REGISTER_TA_STORE`. Doing a detour via `TEE_TS_REGISTER_TS_STORE` will result in a confusing history.

The diff of this PR looks OK though. Please fix up as suggested and I'll go over it once more.

jellesels-arm commented on scripts/ts_bin_to_c.py:
```diff
@@ -59,24 +66,44 @@ def ta_get_flags(ta_f):
+    if args.sp is not None:
+        ts = args.sp
+        is_sp = True
 
```
Done

jellesels-arm commented:
> The last commit "core: create separate stores for SPs and TAs" is really a fixup commit which should be merged into the the earlier commits as appropriate.
> 
> I'd rather go directly from `TEE_TA_REGISTER_TA_STORE` to `TEE_TS_REGISTER_TA_STORE`. Doing a detour via `TEE_TS_REGISTER_TS_STORE` will result in a confusing history.
> 
> The diff of this PR looks OK though. Please fix up as suggested and I'll go over it once more.

done

jenswi-linaro reviewed (commented):
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jellesels-arm commented on core/sub.mk:
```diff
@@ -24,15 +24,27 @@ define process_early_ta
 	process_early_ta,$(out-dir)/ta/$(f).stripped.elf)))
 endif
 
+ifeq ($(CFG_WITH_SP)-$(CFG_EMBEDDED_SP),y-y)
```
We do need to check for CFG_WITH_SP.  Only CFG_EMBEDDED_SP is automatically set. CFG_WITH_SP is used to enable/disable SPs

jenswi-linaro replied:
Wouldn't it make sense to enable `CFG_WITH_SP` automatically with `EMBEDDED_SP_PATHS=y`?
This would also be in line with how `EARLY_TA_PATHS` or `CFG_IN_TREE_EARLY_TAS` enables `CFG_EARLY_TA`

jellesels-arm replied:
Ok, Changed it

jenswi-linaro replied:
Thanks!

jenswi-linaro commented:
Please squash and apply my tag:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
I noticed a couple of spelling mistakes in the commit descriptions. I'm wondering why Travis (checkpatch) did not complain.

jellesels-arm commented:
> I noticed a couple of spelling mistakes in the commit descriptions. I'm wondering why Travis (checkpatch) did not complain.
I fixed 2 of them. I hope that's all of them.

etienne-lms commented on core/arch/arm/kernel/early_ta.c:
```diff
@@ -1,20 +1,21 @@
 #include <kernel/early_ta.h>
 #include <kernel/user_ta.h>
-#include <kernel/user_ta_store.h>
+#include <kernel/ts_store.h>
```
move 1 line up.

jellesels-arm replied:
Done

etienne-lms commented on core/arch/arm/kernel/secstor_ta.c:
```diff
@@ -5,11 +5,11 @@
 #include <tee/tadb.h>
 #include <kernel/user_ta.h>
-#include <kernel/user_ta_store.h>
+#include <kernel/ts_store.h>
```
move 1 line up. here and few other source files.

jellesels-arm replied:
Done

etienne-lms reviewed (commented):
Comments for commit "core: rename ta_store to ts_store".
Otherwise `Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`.

etienne-lms commented on core/arch/arm/include/kernel/embedded_ts.h:
```diff
@@ -0,0 +1,24 @@
+	const uint8_t *ts; /* @size bytes */
+};
+
+#endif /* KERNEL_EARLY_TA_H */
```
s/EARLY/EMBEDDED/
Maybe other _early_ta_ refs should be _embedded_ta_ refs but that's another story.

jellesels-arm replied:
done

jforissier replied:
The comment needs updating: /* KERNEL_EMBEDDED_TS_H */

etienne-lms commented on core/arch/arm/kernel/embedded_ts.c:
```diff
@@ -0,0 +1,216 @@
+static bool decompression_init(z_stream *strm,
+			       const struct embedded_ts *ts)
+{
+	int st;
```
`= Z_OK;`

jellesels-arm replied:
done

etienne-lms commented on core/arch/arm/kernel/embedded_ts.c:
```diff
@@ -0,0 +1,216 @@
+		       const struct embedded_ts* (*find_ts)
+		       (const TEE_UUID *uuid))
+{
+	struct ts_store_handle *handle;
```
init value for these.

jellesels-arm replied:
done

etienne-lms commented on core/arch/arm/kernel/embedded_ts.c:
```diff
@@ -0,0 +1,216 @@
+TEE_Result emb_ts_open(const TEE_UUID *uuid,
+		       struct ts_store_handle **h,
+		       const struct embedded_ts* (*find_ts)
+		       (const TEE_UUID *uuid))
```
maybe prefer
```c
		       const struct embedded_ts*
		       (*find_ts)(const TEE_UUID *uuid))
```

jellesels-arm replied:
done

etienne-lms commented on core/arch/arm/kernel/embedded_ts.c:
```diff
@@ -0,0 +1,216 @@
+
+	if (ts->uncompressed_size) {
+		st = decompression_init(&handle->strm, ts);
+		if (!st) {
```
suggesting
```c
		if (!decompression_init(&handle->strm, ts)) {
			free(handle);
			return TEE_ERROR_BAD_FORMAT;
		}
```
Prevents having 2 local vars names `ts` and `st`.

jellesels-arm replied:
done

etienne-lms commented on core/arch/arm/kernel/embedded_ts.c:
```diff
@@ -0,0 +1,216 @@
+	z_stream *strm = &h->strm;
+	size_t total = 0;
+	uint8_t *tmpbuf = NULL;
+	TEE_Result ret;
```
init vars

etienne-lms commented on core/arch/arm/kernel/embedded_ts.c:
```diff
@@ -0,0 +1,216 @@
+			strm->avail_out = MIN(len - total, 1024U);
+		}
+	} while ((st == Z_OK || st == Z_BUF_ERROR) && (total != len));
+	if (st != Z_OK && st != Z_STREAM_END) {
```
an empty line above?

jellesels-arm replied:
done

jellesels-arm replied:
done

etienne-lms reviewed (commented):
Commit "core: move early_ta implementation to embedded_ts":  missing local vars init (A-b tag once addressed + some suggestions.

Commit "core: create embedded_ts": `KERNEL_EARLY_TA_H` to fix.
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`.

etienne-lms commented on core/arch/arm/kernel/embedded_sp.c:
```diff
@@ -0,0 +1,61 @@
+
+static const struct embedded_ts *find_embedded_sp(const TEE_UUID *uuid)
+{
+	const struct embedded_ts *sp;
```
`= NULL;`

etienne-lms commented on core/arch/arm/kernel/embedded_sp.c:
```diff
@@ -0,0 +1,61 @@
+
+static TEE_Result embedded_sp_init(void)
+{
+	const struct embedded_ts *sp;
```
`= NULL;`

jellesels-arm replied:
done

etienne-lms commented on core/arch/arm/kernel/sub.mk:
```diff
@@ -2,9 +2,13 @@ ifeq ($(CFG_WITH_USER_TA),y)
 endif
+
+srcs-$(CFG_EMBEDDED_SP) += embedded_sp.c
+
```
could remove these empty lines.

jellesels-arm replied:
done

etienne-lms commented on core/arch/arm/include/kernel/embedded_sp.h:
```diff
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2017, Linaro Limited
```
`2020, Arm` ?

jellesels-arm replied:
done

etienne-lms reviewed (changes requested):
commit "core: add embedded SPs": missing few local vars init, 
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` + suggestions.

etienne-lms commented on core/arch/arm/kernel/early_ta.c:
```diff
@@ -16,14 +16,14 @@
+static const struct embedded_ts *find_early_ta(const TEE_UUID *uuid)
 {
-	const struct early_ta *ta;
+	const struct embedded_ts *ta;
```
maybe also add an init value. (could be made in a later P-R) (here and occurrences below)

jellesels-arm replied:
done

etienne-lms reviewed (commented):
commit "core: create embedded_ts"

etienne-lms commented on core/arch/arm/kernel/early_ta.c:
```diff
@@ -6,21 +6,17 @@
+#include <kernel/embedded_ts.h>
 #include <kernel/ts_store.h>
 #include <kernel/user_ta.h>
+#include <kernel/user_ta.h>
```
typo :)

jellesels-arm replied:
done

etienne-lms commented on core/arch/arm/include/kernel/embedded_sp.h:
```diff
@@ -0,0 +1,16 @@
+#define for_each_embedded_sp(_sp) \
+	SCATTERED_ARRAY_FOREACH(_sp, embedded_sps, struct embedded_ts)
+
+#endif /* KERNEL_EMBEDDED_SP_H */
```
minor: thie change belongs to commit "core: create embedded_ts".

jellesels-arm replied:
embedded_sp.h is only created in core: add embedded SPs

etienne-lms replied:
current commit "core/arch/arm/include/kernel/embedded_ts.h" creates **embedded_sp.h**.
then s/`KERNEL_EARLY_SP_H`/`KERNEL_EMBEDDED_SP_H`/ is currently fixed in later commit "core: add embedded SPs".

etienne-lms commented:
Checkpatch (travis) caught:
```shell
8a720d04 core: create embedded_ts
WARNING: space prohibited before semicolon
#123: FILE: core/arch/arm/kernel/early_ta.c:68:
+	const struct embedded_ts *ta = NULL ;
```
IBART error is unclear. Seems to halt in SDP tests. a timeout? a hang? related to this change?

jforissier commented on core/include/kernel/ts_store.h:
```diff
@@ -0,0 +1,74 @@
+			   struct ts_store_handle **h);
+	/*
+	 * Return the size of the unencrypted TS binary, that is: the TS
+	 * header (struct ts_head) plus the ELF data.
```
`struct ts_head` doesn't exist at this point

jforissier commented on core/include/kernel/ts_store.h:
```diff
@@ -0,0 +1,74 @@
+			      uint8_t *tag, unsigned int *tag_len);
+	/*
+	 * Read the TS sequentially, from the start of the TS header (struct
+	 * ts_head) up to the end of the ELF.
```
`struct ts_head` doesn't exist at this point

jforissier commented on core/include/kernel/ts_store.h:
```diff
@@ -0,0 +1,74 @@
+ * Also note that TS storage is sorted lexicographically instead of
+ * numerically.
+ */
+#define TEE_TS_REGISTER_TA_STORE(prio) \
```
We could do without the `TEE_TS_` prefix perhaps

jforissier commented on core/include/kernel/ts_store.h:
```diff
@@ -59,6 +59,9 @@ struct ts_store_ops {
  *
+ * Separate stores are created for TAs and SPs. The user of the stores
+ * knows if it needs to access the TA store or if it needs to access the SP one.
+ *
```
Remove this part and add a comment specific to `TEE_TS_REGISTER_SP_STORE` instead.

jforissier commented on core/include/kernel/ts_store.h:
```diff
@@ -71,4 +74,8 @@ struct ts_store_ops {
 	SCATTERED_ARRAY_DEFINE_PG_ITEM_ORDERED(ta_stores, prio, \
 					       struct ts_store_ops)
 
+#define TEE_TS_REGISTER_SP_STORE(prio) \
```
Here also the prefix doesn' t seem super useful

jforissier reviewed (commented):
A few comments below. Also, I think a PR is needed for the documentation to define TSs, SPs, embedded SP and their relationship with TAs and early TAs.

TBH I would have preferred "early" instead of "embedded" since the concept of embedded TSs is so similar to early TAs. When I chose the name "early TA" I had considered "embedded" too. But not only was "early" shorter and easier to read, it also carried the information that these TAs are not just linked with the TEE but they also available earlier than other TAs. This is important because some services that TAs may depend on may not be available immediately (`tee-supplicant` and therefore secure storage). In other words, not all TAs are eligible for being linked in as an early TA. I suppose it would be the same for SPs?

jellesels-arm commented:
> Checkpatch (travis) caught:
> 
> ```shell
> 8a720d04 core: create embedded_ts
> WARNING: space prohibited before semicolon
> #123: FILE: core/arch/arm/kernel/early_ta.c:68:
> +	const struct embedded_ts *ta = NULL ;
> ```
> 
Done 
> IBART error is unclear. Seems to halt in SDP tests. a timeout? a hang? related to this change?





> A few comments below. Also, I think a PR is needed for the documentation to define TSs, SPs, embedded SP and their relationship with TAs and early TAs.
I was planning to add the documentation when adding the Secure partitions themselves. 
> 
> TBH I would have preferred "early" instead of "embedded" since the concept of embedded TSs is so similar to early TAs. When I chose the name "early TA" I had considered "embedded" too. But not only was "early" shorter and easier to read, it also carried the information that these TAs are not just linked with the TEE but they also available earlier than other TAs. This is important because some services that TAs may depend on may not be available immediately (`tee-supplicant` and therefore secure storage). In other words, not all TAs are eligible for being linked in as an early TA. I suppose it would be the same for SPs?

The reasons, I didn't go for early_sp is cause SP are always loaded as part of the boot process. There is no possibility to load a SP at a latter stage. If we would call it early_sp that it would imply these SPs are specifically loaded during boot and that there are some other SPs that can be loaded later. 
We are also still keeping the naming for the early_ta's. We are saying that embedded_ts are trusted services which are linked inside the image. With early_ta's being a subset of this. With the distinguish it being a TA and it's run early.

jforissier commented:
> The reasons, I didn't go for early_sp is cause SP are always loaded as part of the boot process. There is no possibility to load a SP at a latter stage.

Then I think `embedded_sp` should be called `secure_partition`.

jellesels-arm commented:
> > The reasons, I didn't go for early_sp is cause SP are always loaded as part of the boot process. There is no possibility to load a SP at a latter stage.
> 
> Then I think `embedded_sp` should be called `secure_partition`.
I was planning to add a separate secure_partition which would contain the logic needed for the secure partitions. 
Than we would have embedded_sp deals with how the secure partitions are stored inside the image and secure_partition to deal with the logic. I can combine the two of them in the same file and not have a separation between loading them and the logic.

jellesels-arm commented:
> > > The reasons, I didn't go for early_sp is cause SP are always loaded as part of the boot process. There is no possibility to load a SP at a latter stage.
> > 
> > 
> > Then I think `embedded_sp` should be called `secure_partition`.
> 
> I was planning to add a separate secure_partition which would contain the logic needed for the secure partitions.
> Than we would have embedded_sp deals with how the secure partitions are stored inside the image and secure_partition to deal with the logic. I can combine the two of them in the same file and not have a separation between loading them and the logic.
> 
> I will change the naming off embedded_sp to secure_partition after we get #4180 merged

Renamed embedded_sp to secure_partition and rebased

jforissier commented:
> Renamed embedded_sp to secure_partition and rebased

Thanks, this all looks good to me. You may add to all commits:

`Acked-by: Jerome Forissier <jerome@forissier.org>`

etienne-lms reviewed (approved):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`
